rebased on top of master

Shippable flake seems unrelated
```

--- FAIL: TestRoundTripTypes (5.77s)
serialization_test.go:72: 1: ThirdPartyResourceList: diff: (*expapi.ThirdPartyResourceList){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ListMeta:(api.ListMeta){SelfLink:(string) ResourceVersion:(string)9331749605378857674} Items:([]expapi.ThirdPartyResource)[{TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)堨ĝȚ GenerateName:(string)ó盞 Namespace:(string) SelfLink:(string)亵ȲbQȶ(ĉʉ UID:(types.UID)ȓ冨d垦褻^ʡ*<Ɔ淺垇ư掿 ResourceVersion:(string)11177000488705484421 Generation:(int64)0 CreationTimestamp:(util.Time){Time:(time.Time){sec:(int64)0 nsec:(int32)0 loc:(*time.Location)<nil>}} DeletionTimestamp:(*util.Time)<nil> DeletionGracePeriodSeconds:(*int64)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Description:(string)pǃ茎 ĉǇ釬ʦų6uqW匎Û Versions:([]expapi.APIVersion)[{Name:(string)騜å皾Ɉ椖t{跁Ʃ APIGroup:(string)
A: }]}]}
B: experimental}]}]}
Codec: &{0xc2080d8810 v1}
Data: {"kind":"ThirdPartyResourceList","apiVersion":"v1","metadata":{"resourceVersion":"9331749605378857674"},"items":[{"metadata":{"name":"堨ĝȚ","generateName":"ó盞","selfLink":"亵ȲbQȶ(ĉʉ","uid":"ȓ冨d垦褻^ʡ*\u003cƆ淺垇ư掿","resourceVersion":"11177000488705484421","creationTimestamp":null},"description":"pǃ茎 ĉǇ釬ʦų6uqW匎Û","versions":[{"name":"騜å皾Ɉ椖t{跁Ʃ"}]}]}
Source: "(*expapi.ThirdPartyResourceList){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ListMeta:(api.ListMeta){SelfLink:(string) ResourceVersion:(string)9331749605378857674} Items:([]expapi.ThirdPartyResource)[{TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)堨ĝȚ GenerateName:(string)ó盞 Namespace:(string) SelfLink:(string)亵ȲbQȶ(ĉʉ UID:(types.UID)ȓ冨d垦褻^ʡ*<Ɔ淺垇ư掿 ResourceVersion:(string)11177000488705484421 Generation:(int64)0 CreationTimestamp:(util.Time){Time:(time.Time){sec:(int64)0 nsec:(int32)0 loc:(*time.Location)<nil>}} DeletionTimestamp:(*util.Time)<nil> DeletionGracePeriodSeconds:(*int64)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Description:(string)pǃ茎 ĉǇ釬ʦų6uqW匎Û Versions:([]expapi.APIVersion)[{Name:(string)騜å皾Ɉ椖t{跁Ʃ APIGroup:(string)}]}]}"
Final: "(*expapi.ThirdPartyResourceList){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ListMeta:(api.ListMeta){SelfLink:(string) ResourceVersion:(string)9331749605378857674} Items:([]expapi.ThirdPartyResource)[{TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)堨ĝȚ GenerateName:(string)ó盞 Namespace:(string) SelfLink:(string)亵ȲbQȶ(ĉʉ UID:(types.UID)ȓ冨d垦褻^ʡ*<Ɔ淺垇ư掿 ResourceVersion:(string)11177000488705484421 Generation:(int64)0 CreationTimestamp:(util.Time){Time:(time.Time){sec:(int64)0 nsec:(int32)0 loc:(*time.Location)<nil>}} DeletionTimestamp:(*util.Time)<nil> DeletionGracePeriodSeconds:(*int64)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Description:(string)pǃ茎 ĉǇ釬ʦų6uqW匎Û Versions:([]expapi.APIVersion)[{Name:(string)騜å皾Ɉ椖t{跁Ʃ APIGroup:(string)experimental}]}]}"
FAIL
FAIL k8s.io/kubernetes/pkg/api 6.687s
```

--- a/pkg/kubectl/resource/builder.go
+++ b/pkg/kubectl/resource/builder.go
@@ -184,6 +184,28 @@ func (b *Builder) ResourceTypes(types ...string) *Builder {
 	return b
 }
 
+// ResourceNames accepts a default type and one or more names, and creates tuples of
+// resources
+func (b *Builder) ResourceNames(resource string, names ...string) *Builder {
+	for _, name := range names {
+		// See if this input string is of type/name format
+		tuple, ok, err := splitResourceTypeName(name)
+		if err != nil {
+			b.errs = append(b.errs, err)
+			return b
+		}
+
+		if ok {
+			b.resourceTuples = append(b.resourceTuples, tuple)
+			continue
+		}
+
+		// Use the given default type to create a resource tuple
+		b.resourceTuples = append(b.resourceTuples, resourceTuple{Resource: resource, Name: name})
+	}
+	return b
+}
+
 // SelectorParam defines a selector that should be applied to the object types to load.
 // This will not affect files loaded from disk or URL. If the parameter is empty it is
 // a no-op - to select all resources invoke `b.Selector(labels.Everything)`.
@@ -279,17 +301,14 @@ func (b *Builder) ResourceTypeOrNameArgs(allowEmptySelector bool, args ...string
 			return b
 		}
 		for _, s := range args {
-			seg := strings.Split(s, "/")
-			if len(seg) != 2 {
-				b.errs = append(b.errs, fmt.Errorf("arguments in resource/name form may not have more than one slash"))
+			tuple, ok, err := splitResourceTypeName(s)
+			if err != nil {
+				b.errs = append(b.errs, err)
 				return b
 			}
-			resource, name := seg[0], seg[1]
-			if len(resource) == 0 || len(name) == 0 || len(SplitResourceArgument(resource)) != 1 {
-				b.errs = append(b.errs, fmt.Errorf("arguments in resource/name form must have a single resource and name"))
-				return b
+			if ok {
+				b.resourceTuples = append(b.resourceTuples, tuple)
 			}
-			b.resourceTuples = append(b.resourceTuples, resourceTuple{Resource: resource, Name: name})
 		}
 		return b
 	}
@@ -344,6 +363,23 @@ func hasCombinedTypeArgs(args []string) (bool, error) {
 	default:
 		return false, nil
 	}
+}
+
+// splitResourceTypeName handles type/name resource formats and returns a resource tuple
+// (empty or not), whether it successfully found one, and an error
+func splitResourceTypeName(s string) (resourceTuple, bool, error) {
+	if !strings.Contains(s, "/") {
+		return resourceTuple{}, false, nil
+	}
+	seg := strings.Split(s, "/")
+	if len(seg) != 2 {
+		return resourceTuple{}, false, fmt.Errorf("arguments in resource/name form may not have more than one slash")
+	}
+	resource, name := seg[0], seg[1]
+	if len(resource) == 0 || len(name) == 0 || len(SplitResourceArgument(resource)) != 1 {
+		return resourceTuple{}, false, fmt.Errorf("arguments in resource/name form must have a single resource and name")
+	}
+	return resourceTuple{Resource: resource, Name: name}, true, nil
 }
 
 // Flatten will convert any objects with a field named "Items" that is an array of runtime.Object

--- a/pkg/kubectl/resource/builder_test.go
+++ b/pkg/kubectl/resource/builder_test.go
@@ -114,6 +114,10 @@ func testData() (*api.PodList, *api.ServiceList) {
 		Items: []api.Service{
 			{
 				ObjectMeta: api.ObjectMeta{Name: "baz", Namespace: "test", ResourceVersion: "12"},
+				Spec: api.ServiceSpec{
+					Type:            "ClusterIP",
+					SessionAffinity: "None",
+				},
 			},
 		},
 	}
@@ -374,7 +378,7 @@ func TestResourceByName(t *testing.T) {
 		t.Fatalf("unexpected response: %v %t %#v", err, singular, test.Infos)
 	}
 	if !reflect.DeepEqual(&pods.Items[0], test.Objects()[0]) {
-		t.Errorf("unexpected object: %#v", test.Objects())
+		t.Errorf("unexpected object: %#v", test.Objects()[0])
 	}
 
 	mapping, err := b.Do().ResourceMapping()
@@ -415,6 +419,34 @@ func TestMultipleResourceByTheSameName(t *testing.T) {
 
 	if _, err := b.Do().ResourceMapping(); err == nil {
 		t.Errorf("unexpected non-error")
+	}
+}
+
+func TestResourceNames(t *testing.T) {
+	pods, svc := testData()
+	b := NewBuilder(latest.RESTMapper, api.Scheme, fakeClientWith("", t, map[string]string{
+		"/namespaces/test/pods/foo":     runtime.EncodeOrDie(latest.Codec, &pods.Items[0]),
+		"/namespaces/test/services/baz": runtime.EncodeOrDie(latest.Codec, &svc.Items[0]),
+	})).
+		NamespaceParam("test")
+
+	test := &testVisitor{}
+
+	if b.Do().Err() == nil {
+		t.Errorf("unexpected non-error")
+	}
+
+	b.ResourceNames("pods", "foo", "services/baz")
+
+	err := b.Do().Visit(test.Handle)
+	if err != nil || len(test.Infos) != 2 {
+		t.Fatalf("unexpected response: %v %#v", err, test.Infos)
+	}
+	if !reflect.DeepEqual(&pods.Items[0], test.Objects()[0]) {
+		t.Errorf("unexpected object: \n%#v, expected: \n%#v", test.Objects()[0], &pods.Items[0])
+	}
+	if !reflect.DeepEqual(&svc.Items[0], test.Objects()[1]) {
+		t.Errorf("unexpected object: \n%#v, expected: \n%#v", test.Objects()[1], &svc.Items[0])
 	}
 }
 

